plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java

```diff
@@ -1842,30 +1861,154 @@ public Optional<ConnectorOutputMetadata> finishInsert(
+        return commitVersion;
+    }
 
-        ColumnMappingMode columnMappingMode = getColumnMappingMode(handle.getMetadataEntry(), handle.getProtocolEntry());
+    private void checkForInsertTransactionConflicts(
+            String queryId,
+            IsolationLevel isolationLevel,
+            DeltaLakeInsertTableHandle insertTableHandle,
+            List<ConnectorTableHandle> sourceTableHandles,
+            DeltaLakeCommitSummary commitSummary,
+            long version,
+            int attemptCount)
+    {
+        checkNoMetadataUpdates(commitSummary);
+        checkNoProtocolUpdates(commitSummary);
+
+        List<TupleDomain<DeltaLakeColumnHandle>> enforcedPartitionConstraints = sourceTableHandles.stream()
+                .filter(sourceTableHandle -> sourceTableHandle instanceof DeltaLakeTableHandle)
+                .map(DeltaLakeTableHandle.class::cast)
+                .filter(tableHandle -> insertTableHandle.getTableName().equals(tableHandle.getSchemaTableName())
+                        // disregard time travel table handles
+                        && tableHandle.getReadVersion() >= insertTableHandle.getReadVersion())
+                .map(DeltaLakeTableHandle::getEnforcedPartitionConstraint)
+                .collect(toImmutableList());
+        if (!enforcedPartitionConstraints.isEmpty()) {
+            checkIfCommittedAddedFilesConflictWithCurrentOperation(TupleDomain.columnWiseUnion(enforcedPartitionConstraints), isolationLevel, commitSummary);
+            checkIfCommittedRemovedFilesConflictWithCurrentOperation(isolationLevel, commitSummary);
+        }
+        LOG.debug("Completed checking for conflicts in the query %s for target table version: %s Attempt: %s ", queryId, version, attemptCount);
+    }
+
+    private static void checkNoProtocolUpdates(DeltaLakeCommitSummary commitSummary)
+    {
+        if (commitSummary.getProtocol().isPresent()) {
+            throw new TransactionFailedException("Conflicting concurrent writes found. Protocol changed by concurrent write operation");
+        }
+    }
+
+    private static void checkNoMetadataUpdates(DeltaLakeCommitSummary commitSummary)
+    {
+        if (!commitSummary.getMetadataUpdates().isEmpty()) {
+            throw new TransactionFailedException("Conflicting concurrent writes found. Metadata changed by concurrent write operation");
+        }
+    }
+
+    private static void checkIfCommittedAddedFilesConflictWithCurrentOperation(TupleDomain<DeltaLakeColumnHandle> enforcedPartitionConstraint, IsolationLevel isolationLevel, DeltaLakeCommitSummary commitSummary)
+    {
+        List<AddFileEntry> addedFiles = switch (isolationLevel) {
+            case WRITESERIALIZABLE -> commitSummary.getIsBlindAppend().orElse(false)
+                    ? List.of() // Do not conflict with blind appends
+                    : commitSummary.getAddedFiles();
+            case SERIALIZABLE -> commitSummary.getAddedFiles();
```

Comment:
There should be no conflicts possible when doing blind appends in either situation, right?